File: week-2/word-search-ii/word-search-ii.js
/**
 * @param {character[][]} board
 * @param {string[]} words
 * @return {string[]}
 */
var findWords = function(board, words) {
    const trie = {};
    for (let word of words) {
        let ptr = trie;
        for (let letter of word) {
            if (!(letter in ptr)) ptr[letter] = {};
            ptr = ptr[letter];
        }
        ptr['@'] = true;
    }
    
    const res = [];
    for (i in board) {
        for (j in board[i]) {
            dfs(parseInt(i), parseInt(j), board, '', trie, res);
        }
    }
    return res;
};

const dfs = (i, j, board, word, trie, res) => {
    if ('@' in trie) {
        res.push(word);
        delete trie['@'];
    }
    
    if (i<0 || j<0 || i>board.length-1 || j>board[0].length-1) return;
    
    const letter = board[i][j];
    if (board[i][j] !== '#' && (letter in trie)) {
        board[i][j] = '#';
        dfs(i+1, j, board, word + letter, trie[letter], res);
        dfs(i-1, j, board, word + letter, trie[letter], res);
        dfs(i, j+1, board, word + letter, trie[letter], res);
        dfs(i, j-1, board, word + letter, trie[letter], res);
        board[i][j] = letter;
    }
    
    return;
}

File: week-2/word-search-ii/word-search-ii.py
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = {}
        for word in words:
            ptr = trie
            for letter in word:
                if letter not in ptr:
                    ptr[letter] = {}
                ptr = ptr[letter]
            ptr['@'] = True
        
        res = []
        for i in range(len(board)):
            for j in range(len(board[i])):
                self.dfs(i, j, trie, '', board, res)
                
        return res
        
    def dfs(self, i, j, trie, word, board, res):
        if '@' in trie:
            res.append(word)
            del trie['@']
            
        if i<0 or j<0 or i>len(board)-1 or j>len(board[0])-1:
            return
        
        letter = board[i][j]
        if board[i][j] != '#' and board[i][j] in trie:
            board[i][j] = '#'
            self.dfs(i-1, j, trie[letter], word + letter, board, res)
            self.dfs(i+1, j, trie[letter], word + letter, board, res)
            self.dfs(i, j-1, trie[letter], word + letter, board, res)
            self.dfs(i, j+1, trie[letter], word + letter, board, res)
            board[i][j] = letter
        
        return
